### DISCUSSION TITLE
How to use Alpine's plugin system with a custom Filament package?

### DISCUSSION THREAD

**Discussion by RVxLab** [filamentphp/filament#1187] [Help] [answered]
I've been working on getting my colour picker package to work with Filament 2.x for the last couple of days but I'm currently stuck on getting my Blade component to make use of it.

Taking inspiration of the DateTimePicker component I attempt to create a custom plugin that registers my plugin in Alpine.data as such:

```js
function colorPickerPlugin(Alpine) {
    console.log(Alpine);

    Alpine.data('rvxlabFilamentColorPickerComponent', () => ({
        init() {
            console.log('init');
        },
    }));
}

window.Alpine.plugin(colorPickerPlugin);
```

In this case I do get the Alpine object printed to console, so the JS is running.

However, Alpine does not seem to be able to find the `rvxlabFilamentColorPickerComponent` at all, giving an Alpine expression error.
I also tried to wrap this in an alpine:init listener, but that one is never called. 

Could anyone tell me what I'm doing wrong here? I just can't seem to figure it out.

Thanks.

Commit: https://github.com/RVxLab/filament-colorpicker/commit/877e5d215f147ada7c7a23baa9db6984730649b5

**Comment by danharrin** [comment, marked as answer]
Hi, have you tried doing this in an external JS file? And then registering this in your plugin service provider? It may be a weird inconsistency between NPM & browser Alpine builds.

**Reply by RVxLab** [reply]
You mean installing my own Alpine build? If so then no I haven't considered that yet.

It may sound like a sub-optimal solution as you're essentially using 2 different Alpine versions at that point, but it's worth a shot. I'll get back to you with my findings.

**Reply by danharrin** [reply]
No, you shouldn't include 2 Alpine builds. I was just suggesting that you should add the plugin JS to a separate script instead of in a `<script>` (which I assume you're using now, right?)

**Reply by danharrin** [reply]
Here's an example of someone who's added a Alpine component to a Filament 2 plugin - https://github.com/humaidem/filament-map-picker/blob/main/resources/js/map-picker.js

**Reply by RVxLab** [reply]
> No, you shouldn't include 2 Alpine builds. I was just suggesting that you should add the plugin JS to a separate script instead of in a <script> (which I assume you're using now, right?)

I am now in the current development version.

> Here's an example of someone who's added a Alpine component to a Filament 2 plugin - https://github.com/humaidem/filament-map-picker/blob/main/resources/js/map-picker.js

Oooh, thanks! I'll take a look at this and see if that works.

**Reply by RVxLab** [reply]
Taking the map picker as inspiration worked really well.

I take now take the initial colour as starting point for the colour picker and then run `$wire.set` with the defer flag set to update the underlying component. The rest I handle in JavaScript.

Thanks again for the pointer to that, it was exactly what I needed.

**Reply by danharrin** [reply]
Great, let me know when you publish v2 and I'll add it to our plugin list on Discord.

**Reply by RVxLab** [reply]
I just tagged v1.0.0 of the package.

Thanks again for your help!

**Reply by danharrin** [reply]
Great! Thanks for contributing to our community

**Reply by danharrin** [reply]
By the way - `PluginServiceProvider` extends `PackageServiceProvider` :D

**Reply by RVxLab** [reply]
> Great! Thanks for contributing to our community

And thank you for making such an awesome project

> By the way - `PluginServiceProvider` extends `PackageServiceProvider` :D

Nice! I'll merge the service providers soon then!
